src/economy/economy.cpp

```diff
@@ -519,64 +546,120 @@ float artisan_scale_limit(sys::state& state, dcon::nation_id n, dcon::commodity_
-				//auto limit = artisan_scale_limit(state, n, cid);
-				auto profit = base_artisan_profit(state, n, cid); //* limit;
-				profits[cid.index()] = profit;
+	return f;
+}
 
-				if(profit > 0.f) {
-					total_weights += profit;
-					total_possible += 1.0f;
-				}
-			}
+float get_artisans_multiplier(sys::state& state, dcon::nation_id n) {
+	float multiplier = 0.000001f * state.world.nation_get_everyday_needs_costs(n, state.culture_definitions.artisans);
+	return 1.f / (multiplier + 1.f);
+}
+
+void scores_to_distribution(std::vector<float> & scores, std::vector<float> & distribution, float multiplier) {
+	float max_score = std::numeric_limits<float>::min();
+	for(uint32_t i = 1; i < scores.size(); ++i) {
+		if(scores[i] > max_score) {
+			max_score = scores[i];
 		}
+	}
 
-		for(uint32_t i = 1; i < csize; ++i) {
-			dcon::commodity_id cid{ dcon::commodity_id::value_base_t(i) };
-			auto kf = state.world.commodity_get_key_factory(cid);
+	float total = 0.f;	
 
-			float market_size = state.world.commodity_get_total_consumption(cid) * 0.001f;
-			auto consumed_ratio = std::min(1.f, (state.world.commodity_get_total_consumption(cid) + 0.0001f) / (state.world.commodity_get_total_production(cid) + 0.0001f));
-			float output_size = state.world.commodity_get_artisan_output_amount(cid);
+	// crude approximation of softmax
+	for(uint32_t i = 1; i < scores.size(); ++i) {
+		distribution[i] = scores[i] - max_score;
+		distribution[i] = pseudo_exp_for_negative(distribution[i] * multiplier);
+		total += distribution[i];
+	}
 
-			if(state.world.commodity_get_artisan_output_amount(cid) > 0.0f && (state.world.commodity_get_is_available_from_start(cid) || (kf && state.world.nation_get_active_building(n, kf)))) {
-				auto& w = state.world.nation_get_artisan_distribution(n, cid);
-				if(profits[cid.index()] < 0) {
-					w = std::max(0.f, (1.0f + profits[cid.index()] * w * market_size / consumed_ratio / consumed_ratio / output_size - 100.f / num_artisans) * w);
-				} else if(total_weights > 0.0001f) {
-					float ideal_weight = (profits[cid.index()] / total_weights);
-					float speed = std::clamp(distribution_drift_speed * w * profits[cid.index()] * market_size / output_size + 100.f / num_artisans, 0.f, 0.01f);
-					w = std::clamp((1.0f - speed) * w + speed * ideal_weight, 0.f, 1.f);
-
-					assert(std::isfinite(w));
-				} else if (total_possible > 0.f) {
-					// move toward 0 production
-					w = (1.0f - distribution_drift_speed) * w;
-					assert(std::isfinite(w));
-				}
-			} else {
-				state.world.nation_set_artisan_distribution(n, cid, 0.0f);
-			}
+	for(uint32_t i = 1; i < scores.size(); ++i) {
+		distribution[i] = distribution[i] / (total + 0.001f);
+	}
+}
+
+void fill_artisan_distribution(sys::state& state, dcon::nation_id n, std::vector<float>& distribution) {
+	std::vector<float> scores;
```

Comment:
so, there is a minor inefficiency here. Going to the memory manager to dynamically allocate a buffer is better avoided if you don't need it. In this case, the only reason you seem to be storing them at all is to calculate the maximum. Since the calculations are simple enough, it is almost certainly faster to loop over all the commodities and find the maximum and then calculate each score again as you need it, thereby saving the need to dynamically allocate an array at all.